<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;"> </center> 

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
# Импортируем библиотеки для работы
import pandas as pd
import psycopg2
import urllib.parse as req
import requests
from bs4 import BeautifulSoup
import plotly.express as px

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
select
   count(id)
from vacancies
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
select
   count(id)
from public.EMPLOYERS
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
select
   count(id)
from public.areas
'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
select
   count(id)
from public.industries
'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

### выводы по предварительному анализу данных

В рамках проекта представлено 4 уникальные таблицы: 
* Vacancies - данные с вакансиями, содержат 49197 строк;
* Areas - справочник с регионами, содержат 1362 строк;
* Employers - справчник с наименованием работадателей, содержат 23501 строк;
* Industries - справчник со сферами деятельности, содержат 294 строк. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''
SELECT 
    a.name area, 
    COUNT(*) cnt
FROM areas a
    JOIN VACANCIES v on a.id = v.area_id 
GROUP BY 1
ORDER BY 2 DESC
LIMIT 5
'''

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''
select
    count(*)
from vacancies
where salary_from is not null
    or salary_to is not null
'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''
select
    round(avg(salary_from), 0) avg_from,
    round(avg(salary_to), 0) avg_to
from vacancies
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_from    avg_to
0   71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''
select
   schedule,
   employment,
   count(*) cnt
from vacancies
group by 1, 2
order by 3 desc
offset 1 limit 1
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule        employment   cnt
0  Удаленная работа  Полная занятость  7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''
select
   experience as "Требуемый опыт работы",
   count(*) cnt
from vacancies
group by 1
order by 2
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Требуемый опыт работы    cnt
0           Более 6 лет   1337
1             Нет опыта   7197
2         От 3 до 6 лет  14511
3    От 1 года до 3 лет  26152

***

## выводы по детальному анализу вакансий

После получения первичных данных с первого блока, переходим к анализу вакансий. 

* В первом задании определили количество вакансий в каждом регионе и как ожидалось, чем больше регион по численностью проживающих людей, тем больше вакансий размещено. Соответственно  на первом месте это Москва (5333 вакансий), следующим стал Санкт-Петербург (2851) и далее города идут по убыванию;
* Во-втором задании определили количество вакансий с представленной вилкой по зарплате либо по верхнему пределу, либо по нижнему пределу, или наличие двух вилок одновременно. Данный показатель составляет 24073 вакансии. Если вспомнить, что всего вакансий 49197, то не трудно заменить, что только половину вакансий имеют вилку. Соответственно вакансий с пустыми вилками составляет (49197-24073) 25124. Честно говоря, мне казалось, что пустых вилок будет намного больше. Бытует мнение, что это делается для того, чтобы при собеседовании с кандидатом тот сам озвучивал желаемую зарплату и если озвученная зарплата будет выше той суммы, что заложена в штатном расписание, кандидату просто вежливо об этом скажут, а если кандидат назовет предпочитаемую зарплату меньше, чем указанная в штатном расписание, то компания скорей всего согласиться на эти условия так как она потенциально сэкономит на человеке. Этот трюк делается лишь потому, что ФОТ составляет большую или одну из больших статей затрат в любой организации;
*   В-третьем задании мы выявили средние показатели вилок по зарплате нижнего предела (71065 руб.) и верхнего предела (110537 руб.) Эти показатели дают представление о потенциальных возможностях для соискателей. Но, сам подход вывода среднего показателя кажется немного сомнительным, потому что:
    1. Согласно вводным данным все зарплаты переведены в рубли - следовательно курс валют на момент перевода мог оказаться выбросом. Так как нет даты, на какое число представлен датасет;
    2. Количество выложенных вакансий в городах с проживающим большем количеством людей больше, чем в городах с меньшим количеством людей, от сюда следует, что из-за конкуренции в мегаполисе по типу Москва и Санкт-Петербург нижний предел зарплат высокий, а при анализе среднего показателя по всем вакансиям среднее число размывается;
    3. Сомнительное число верхнего предела 110537 руб. Данное сомнение заключается в следующем:
        * Все мы понимаем, что на ответственные руководящие должности приглашаются в основном по знакомству и их зарплаты обсуждаются за закрытыми дверями, поэтому таких вакансий не встретить на портале HH.RU;
        * В заключение данного пункта, хочется подчеркнуть, что выявленный расчет условный и на него нельзя оперироваться, без предварительной обработки вакансий, далее для более четкого представления необходимо добавлять градации по регионам и еще по направлениям, так как одна должность ценится больше по отношению к другой должности. 
* В-четвертом задании было выявлено количество вакансий для типа работы и способа работы. Данный показатель является полным и его можно брать за основу при анализе вакансий. Расчет имеет глубинное представление, для примера привожу три первых значений по количеству выложенных вакансий. На первой строке это полный день и полная занятость имеет 35367 вакансий, на второй строке удаленная работа и полный день имеет 7802 вакансии и на третьей строке гибкий график и полная занятость имеет 1593 вакансий. Из этих трех представленных строк можно четко отследить тенденцию в датасете, так основную часть вакансий имеет традиционный способ работы, но с развитием интернет-технологий начинает увеличиваться количество вакансий на удаленном формате работы. Удаленный формат работы будет и дальше развиваться, так как данный формат потенциально может экономить организации деньги на содержание сотрудников в офисе. Третью и остальные строки составляют только (49197-35367-7802) 6028 вакансий. Почему так? Все дело в том, что портал HH.RU больше акцентирован на поиск среднего класса сотрудников, так как такие должности как разнорабочие и грузчики в большей своей массе не пользуется данным порталом, а высшее руководство как было подчеркнуто выше набираются по знакомству и рекомендациям;
* В-пятом задании было рассчитано количество выложенных вакансий по требуемому опыту. Полученные данные является весьма закономерны и нормальные. Так, например, 26152 вакансий имеют тип требуемого опыта "От 1 года до 3 лет", это говорит о том, что чуть больше половины от всех вакансий требуется не большой опыт, скорей всего это связано с тем, что у линейных руководителей бывают несколько штатных сотрудников. Второй тип требуемого опыта "От 3 до 6 лет", что составляет 14511 вакансий следуя моей логике, это скорее уже как раз вакансии линейных руководителей. Третий тип "Нет опыта" и 7197 вакансий, данный тип встречается во всех должностях без исключения, так как это один из способов привлечь людей. Ну и последний тип "Более 6 лет", что составляет 1337 вакансий и тут могут быть как руководители организации, так и суперквалифицированные, но при этом узконаправленные специалисты. 

>* Проведенный детальный анализ данных позволяет понять основные тенденции на рынке труда, но требует дополнительных исследований и обработки данных для получения более точных и релевантных результатов. Также важно учитывать, что различные регионы и сферы деятельности могут иметь существенные отличия, которые необходимо учитывать при анализе. Переходим к следующему разделу посвященный анализу работодателей. 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''
select
   e.name,
   count(*)
from vacancies v
   join employers e on v.employer_id = e.id 
group by 1 
order by 2 desc
limit 5
'''

In [23]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = f'''
SELECT 
    a.name city,
    COUNT(DISTINCT e.id) empl_cnt,
    COUNT(distinct v.id) vacan_cnt
FROM areas a 
    LEFT JOIN employers e ON a.id = e.area
    LEFT JOIN vacancies v ON a.id = v.area_id
where v.id is null
GROUP BY 1
ORDER BY 2 DESC
LIMIT 1 
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


city  empl_cnt  vacan_cnt
0  Россия       410          0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = f'''
SELECT
    e.name,
    COUNT(DISTINCT v.area_id)
FROM employers e
    JOIN vacancies v on e.id = v.employer_id
GROUP BY 1 
ORDER BY 2 DESC
LIMIT 1  
'''

In [27]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0  Яндекс    181

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = f'''
select 
    count(*)
from employers e
    left join employers_industries ei on e.id = ei.employer_id
where ei.industry_id is null 
'''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5 = f'''
select 
    e.name
from employers e
    left join employers_industries ei on e.id = ei.employer_id
GROUP BY 1 
having COUNT(ei.industry_id) = 4 
ORDER BY 1 
offset 2 LIMIT 1
'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''
SELECT
    COUNT(e.name)
FROM industries i
    LEFT JOIN employers_industries ei on ei.industry_id = i.id 
    LEFT JOIN employers e on ei.employer_id = e.id
where i.name = 'Разработка программного обеспечения'
'''

In [33]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников
path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
cities = tuple(pd.read_html(url)[1]['Город'])

In [35]:
# текст запроса
query_5_7 = f'''with city_wiki AS ( 
select
    a.name city_name,
    count(v.id) vac_cnt 
from vacancies v 
    inner join areas a on v.area_id = a.id 
    inner join employers e on v.employer_id = e.id 
WHERE e.name = 'Яндекс' 
    and a.name in {cities}
group by 1 
order by 2 
)
select
    cw.city_name,
    cw.vac_cnt
from city_wiki cw
 
union all 

select
    'Total' city_name,
    sum(vac_cnt) total
from city_wiki
'''

In [36]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


city_name  vac_cnt
0              Омск     21.0
1         Челябинск     22.0
2        Красноярск     23.0
3         Волгоград     24.0
4             Пермь     25.0
5            Казань     25.0
6    Ростов-на-Дону     25.0
7               Уфа     26.0
8            Самара     26.0
9         Краснодар     30.0
10          Воронеж     32.0
11      Новосибирск     35.0
12  Нижний Новгород     36.0
13     Екатеринбург     39.0
14  Санкт-Петербург     42.0
15           Москва     54.0
16            Total    485.0

***

## выводы по анализу работодателей
Проанализировав блок по вакансиям, мы переходим к анализу работодателей. 
* В первом же задании мы выводим наименование организации и количество выложенных вакансий. Полученный результат является весьма ошеломительным т.к. на первом месте находится "Яндекс" с количеством выложенных вакансий 1933, на пятом месте "Газпром нефть" с 331 вакансией. У "Яндекса" такое большое количество вакансий т.к. компания диверсифицирована и род ее большого количества направлений требует большого количества сотрудников, такие например как курьеры, таксисты, операторы кол-центров и т.п. На пятом месте "Газпром нефть" и в основном там нужны люди в большом количестве для обслуживания нефтеперерабатывающих предприятий, но из-за того, что предприятий не такое большое количество как у "Яндекса", соответственно и людей нужно меньше;
* Во втором задании необходимо было вывести компанию и количество вакансий в разрезе региона, а в ответе указать тот регион, в котором есть большое количество работодателей, но при этом нет вакансий и этот регион стал "Россия". Что тут можно сказать? Судя по ответу можно лишь подчеркнуть, что в датасете представлено 410 компаний находящиеся в "России";
* В третьем задании необходимо было посчитать количество регионов, в которых размещает каждая отдельно взятая компания и естественно на первом месте находится компания "Яндекс" с показателем 181, это лишний раз подтверждает масштаб организации; 
* В четвёртом задании необходимо было высчитать число организаций не указавшие свою сферу деятельности и их оказалось 8419 компаний, это не много так как всего компаний представленные в датасете составляет 40752 организации;
* В пятом задании надо было посчитать количество сфер деятельности у каждой компании и вывести те компании, у которых указано 4 сферы деятельности, а в ответе указать компанию, которая находится в алфавитном порядке на 3 месте и этой компанией оказалась "2ГИС" - данная компания занимается: 
    1. Производство мультимедиа, контента, редакторская деятельность;
    2. Консалтинговые услуги;
    3. Кол-центры, диспетчерские службы;
    4. Интернет-компания (поисковики, платежные системы, соц.сети, информационно-познавательные и развлекательные ресурсы, продвижение сайтов и прочее).
* В шестом задании необходимо было вывести количество работодатель у которых прописана сфера деятельности "Разработка программного обеспечения" и таких компаний оказалось 3553, если учесть, что всего представлено 40752 организации, то это число достаточно хорошее, так как сфер деятельности присутствует огромное множество. Также хочется отметить, что с развитием интернет-технологий количество организаций с такой сферой деятельностью будет только увеличиваться;
* В седьмом задании, надо было взять список городов миллионников из интернета источника "Wikipedia", и посчитать количество выложенных вакансий в конретконкретных населенных пунктах, а также вывести общее количество вакансий и все это сделать напримере организации "Яндекс". Получилось 16 городов и лидерами по количеству выложенных вакансий стали такие города как Москва - 54, Санкт-Петербург - 42 и т.д. а общее количество выложенных вакансий в этих городах составило 485. Данные являются весьма закономерны, так как чем больше город, тем больше компания предоставляет услуг и продуктов, а чтобы предоставлять свои услуги и продукты, нужны как раз эти сотрудники. 

>* Вывод по блоку: Мы подробно ознакомились с компаниями и их сферами деятельности, а также проанализировали количество вакансий с учетом наименований компаний, их сферы деятельности и населенных пунктов, в которых представлены эти вакансии.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
# текст запроса
query_6_1 = f''' 
select
    count(*)
from vacancies
where name ilike '%data%' 
    or name ilike '%данн%'
'''

In [38]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
# текст запроса
query_6_2 = f''' 
select 
    count(*)
from vacancies
where
    (
    lower(name) like '%data scientist%'  or
    lower(name) like '%data science%' or
    lower(name) like '%исследователь данных%'  or
    (name like '%ML%' and name not ilike '%HTML%') or 
    lower(name) like '%machine learning%'  or
    lower(name) like '%машинн%обучен%'
    )
    and
    (
    lower(name) like '%junior%' or
    experience = 'Нет опыта' or
    lower(employment) like '%стажировка%'
    )
'''

In [40]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
# текст запроса
query_6_3 = f''' 
select
    count(*)
from
    vacancies
where
    (
    lower(name) like '%data scientist%'  or
    lower(name) like '%data science%' or
    lower(name) like '%исследователь данных%'  or
    (name like '%ML%' and name not ilike '%HTML%') or 
    lower(name) like '%machine learning%'  or
    lower(name) like '%машинн%обучен%'
    )
    and
    (
    key_skills ilike '%SQL%' or
    key_skills like '%postgres%'
    )
'''

In [42]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
# текст запроса
query_6_4 = f''' 
select
    count(*)
from
    vacancies
where
    (
    lower(name) like '%data scientist%'  or
    lower(name) like '%data science%' or
    lower(name) like '%исследователь данных%'  or
    (name like '%ML%' and name not ilike '%HTML%') or 
    lower(name) like '%machine learning%'  or
    lower(name) like '%машинн%обучен%'
    )
    and
    (key_skills ilike '%Python%')
'''

In [44]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
# текст запроса
query_6_5 = f''' 
select 
    round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2)
from vacancies 
where 
    (
    lower(name) like '%data scientist%'  or
    lower(name) like '%data science%' or
    lower(name) like '%исследователь данных%'  or
    (name like '%ML%' and name not ilike '%HTML%') or 
    lower(name) like '%machine learning%'  or
    lower(name) like '%машинн%обучен%'
    )
'''

In [46]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
# текст запроса
query_6_6 = f''' 
select
    round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0) avg_sal 
from vacancies 
where 
    (
    lower(name) like '%data scientist%'  or
    lower(name) like '%data science%' or
    lower(name) like '%исследователь данных%'  or
    (name like '%ML%' and name not ilike '%HTML%') or 
    lower(name) like '%machine learning%'  or
    lower(name) like '%машинн%обучен%'
    )
    and
    (salary_from is not null or salary_to is not null)
    and 
    experience = 'От 3 до 6 лет'
'''

In [48]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


avg_sal
0  243115.0

***

## выводы по предметному анализу
Изучив предварительно представленные организации  в датасете, мы переходим к самому интересному блоку предметного анализа.

* В первом задании подсчитываем количество вакансий, содержащих определенные ключевые слова в названии, таких вакансий оказалось 1771 штук.;
* Во втором задании исследуем вакансии для начинающих дата-сайентистов с дополнительными условиями, такими как отсутствие опыта, тип работы "стажировка" и наличие слова "Junior" в названии. В итоге находим 51 такую вакансию;
* В третьем задании анализируем вакансии для начинающих дата-сайентистов с требованием навыков "SQL" и "Postgres". Таких вакансий оказалось 201 шт; 
* В четвертом задании надо было посчитать количество вакансий для наинающих дата-сайнтистов, с наличием требуемого знания "Python", таких вакансий окалось 351 шт;
*  В пятом задании надо было вывести средний показатель количества указанных работадателем необходимых навыков для начинающих дата-сайнтистов, этот показатель составил 6.41. Это говорит о том, что нашему специалисту необходимо иметь как минимум 6 навыков, для того, чтобы его смогли взять на работу;
* В шестом задании, необходимо было вывести средний показатель зарплаты дата-сайнтиста имеющий опыт "От 3 до 6 лет". В ответе получилось 243115 руб. Сама по себе зарплата явялется достаточно хорошей и привлекательной, учитывая тот факт, что удаленный тип работы еще увеличивается, то вырисовывается красивая картина. А еще это лишний раз подчеркивает, что программисты и аналитики востребованы на сегодняшнем рынке. 

>* В данном блоке было подробно расмотрен такой специалист как дата-сайнтист. Определили количетсво вакансий с разными условиями, необходимое количетсво нужных навыков и конечно же просчитали среднюю зарплату. 


## 7. Дополнительные исследования данных

Задача найти вакансии в каждом городе для DataScients и определить сколько вакансий имеется и какая там средняя зарплата.
* Для начала выгружаем из БД все вакансии и регионы с фильтром по DS. В ответе получается 480 вакансий;
* Для вычисления средних зарплат мы проводим следующие действия. 
    * Логика заключается в том, что в представленном df_query_7_1 есть salary_from и salary_to. Я сначала делаю сортировка по региону в этом регионе счиатю среднее по salary_from, затем средний показатель salary_to. Когда получая два числа складываю их между собой и делю на два, так я получаю среднюю зарплату в каждом регионе. 
* После первичной обработке в таблице осталось множество пропусков (NaN). Для их обработки выполнены следующие действия: 
    * Сначала заполняем значения в переменной average_df средними зарплатами по регионам: 
    * Затем проходим по average_df и, если в каком-либо регионе нет данных (NaN), ищем значение в df_query_7_1;
    * Если находим число, заменяем NaN на это число, не выполняя дополнительных вычислений;
    * После заполнения данных объединяем average_df с исходным df_query_7_1.
* Теперь необходимо подготовить данные к построению графика. В моем представление график должен из себя представлять график, на котором будут указаны города, количество вакансий в этих городах и средние зарплаты. Для этого понадобиться: 
    * Приводим регионы к общему формату, например, переименовываем названия стран в столицы или переименовываем город Алматы в столицу Астану;
    * После того, как мы получили список городов, необходимо выгрузить широту и долготу городов из интернет-источника для корректного отображения на карте;
    * Создаем график, на котором отображены кружки с местоположением городов, а также цвет кружков, представляющий количество вакансий, и размер кружков, отражающий среднюю зарплату. 

In [49]:
# текст запроса
query_7_1 = f''' 
select
    a.name сity_name,
    v.salary_from,
    v.salary_to,
    v.id,
    v.name vacan_name
from vacancies v
    join areas a on v.area_id = a.id
where 
    (
    lower(v.name) like '%data scientist%'  or
    lower(v.name) like '%data science%' or
    lower(v.name) like '%исследователь данных%'  or
    (v.name like '%ML%' and v.name not ilike '%HTML%') or 
    lower(v.name) like '%machine learning%'  or
    lower(v.name) like '%машинн%обучен%'
    )    
'''

In [50]:
# результат запроса
df_query_7_1 = pd.read_sql_query(query_7_1, connection)

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\408772806.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_query_7_1 = pd.read_sql_query(query_7_1, connection)


#### Для получения средних зарплат необходимо произвести расчет. 

Логика заключается в том, что в представленном df_query_7_1 есть salary_from и salary_to. Я сначала делаю сортировка по региону в этом регионе счиатю среднее по salary_from, затем средний показатель salary_to. Когда получая два числа складываю их между собой и делю на два, так я получаю среднюю зарплату в каждом регионе.

In [51]:
unique_cities = df_query_7_1['сity_name'].unique()
average_salaries = {}

for city in unique_cities:
    city_data = df_query_7_1[df_query_7_1['сity_name'] == city]
    average_salary_from = city_data['salary_from'].sum() / city_data['salary_from'].count()
    average_salary_to = city_data['salary_to'].sum() / city_data['salary_to'].count()
    average_salary = (average_salary_from + average_salary_to) / 2
    average_salaries[city] = average_salary

average_df = pd.DataFrame(list(average_salaries.items()), columns=['сity_name', 'average_salary'])

C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\487738929.py:6: RuntimeWarning: invalid value encountered in scalar divide
  average_salary_from = city_data['salary_from'].sum() / city_data['salary_from'].count()
C:\Users\Георгий\AppData\Local\Temp\ipykernel_11268\487738929.py:7: RuntimeWarning: invalid value encountered in scalar divide
  average_salary_to = city_data['salary_to'].sum() / city_data['salary_to'].count()


#### В таблице пристутсвует множество пропусков и как с этим быть? 

После первичной обработке в таблице осталось множество пропусков (NaN). Для их обработки выполнены следующие действия: 
* Сначала заполняем значения в переменной average_df средними зарплатами по регионам: 
* Затем проходим по average_df и, если в каком-либо регионе нет данных (NaN), ищем значение в df_query_7_1;
    * Если находим число, заменяем NaN на это число, не выполняя дополнительных вычислений;
* После заполнения данных объединяем average_df с исходным df_query_7_1.

In [52]:
#Второй этап
unique_cities = average_df['сity_name'].unique()
aver_salar = {}

for city_name in unique_cities:
    if pd.isna(average_df.loc[average_df['сity_name'] == city_name, 'average_salary'].values[0]):
        city_data = df_query_7_1[df_query_7_1['сity_name'] == city_name]
        if city_data['salary_to'].isna().all() and city_data['salary_from'].isna().all():
            aver_salar[city_name] = None
        else:
            salary_to_avg = city_data['salary_to'].mean()
            salary_from_avg = city_data['salary_from'].mean()
            if pd.notna(salary_to_avg) and pd.notna(salary_from_avg):
                aver_salar[city_name] = (salary_to_avg + salary_from_avg) / 2
            else:
                aver_salar[city_name] = salary_to_avg if pd.notna(salary_to_avg) else salary_from_avg

for city_name, avg_salary in aver_salar.items():
    average_df.loc[average_df['сity_name'] == city_name, 'average_salary'] = avg_salary

average_df = round(average_df,2)

После того, как выведены средние зарплаты, необходимо объединить два df в один

In [53]:
#Объединяю df
merged_df = df_query_7_1.merge(average_df, left_on='сity_name', right_on='сity_name', how='left')


#### Следующий этап создание графика 

* Теперь необходимо подготовить данные к построению графика. В моем представление график должен из себя представлять график, на котором будут указаны города, количество вакансий в этих городах и средние зарплаты. Для этого понадобиться: 
    * Приводим регионы к общему формату, например, переименовываем названия стран в столицы или переименовываем город Алматы в столицу Астану;
    * После того, как мы получили список городов, необходимо выгрузить широту и долготу городов из интернет-источника для корректного отображения на карте;
* Создаем график, на котором отображены кружки с местоположением городов, а также цвет кружков, представляющий количество вакансий, и размер кружков, отражающий среднюю зарплату

In [54]:
# Создаем словарь для переименования городов
rename_dict = {
    "Кипр": "Никосия",
    "Армения": "Ереван",
    "Черногория": "Подгорица",
    "Турция": "Анкара",
    "Сербия": "Белград",
    "Алматы": "Астана"
}

merged_df['сity_name'] = merged_df['сity_name'].replace(rename_dict)

In [55]:
# Открываем сайт и согласно нашим городам импортируем долготу и широту
url = 'https://time-in.ru/coordinates'
response = requests.get(url)

if response.status_code == 200:
    page = BeautifulSoup(response.text, 'html.parser')
    cities = merged_df['сity_name'].tolist()
    city_coordinates = {}
    left_elements = page.find_all('a', class_='coordinates-items-left')
    right_elements = page.find_all('div', class_='coordinates-items-right')

    for left, right in zip(left_elements, right_elements):
        city_name = left.text.strip()
        coordinates = right.text.strip()
        if city_name in cities:
            city_coordinates[city_name] = coordinates

#Разносим долготу и широту по новым столбцам 
merged_df['longitude'] = merged_df['сity_name'].map(lambda x: None if city_coordinates.get(x, None) is None else city_coordinates.get(x, None).split(',')[1])
merged_df['latitude'] = merged_df['сity_name'].map(lambda x: None if city_coordinates.get(x, None) is None else city_coordinates.get(x, None).split(',')[0])



In [56]:
#Предварительно делаю сортировку данных, для отображения графика
pivot_df = pd.pivot_table(merged_df, 
                          index=('сity_name', 'average_salary', 'longitude', 'latitude'), 	
                          values='id', 
                          aggfunc='nunique', 
                          fill_value=0)

pivot_df.sort_values('average_salary', ascending=False)

id
сity_name       average_salary longitude latitude     
Никосия         300000.00       33.36    35.18       5
Ереван          268862.67       44.51    40.18       6
Анкара          233794.00       32.85    39.92       2
Подгорица       233794.00       19.26    42.44       1
Белград         233794.00       20.47    44.8        3
Москва          215375.00       37.62    55.75     217
Санкт-Петербург 183409.60       30.31    59.94      64
Екатеринбург    142833.33       60.61    56.85       7
Рязань          137500.00       39.69    54.63       4
Новосибирск     129833.33       82.93    55.04      23
Астана          94586.50        71.45    51.18      16
Уфа             90000.00        55.97    54.74       2
Владивосток     85000.00        131.87   43.11       3
Пермь           65000.00        56.25    58.01       5
Барнаул         60000.00        83.76    53.36       1
Казань          50000.00        49.12    55.79      16
Ростов-на-Дону  50000.00        39.72    47.23       6

In [57]:
#Задаю параметры графика и отображаю его 
map_df = pivot_df.reset_index()  


fig = px.scatter_geo(map_df,
                     lat='latitude',
                     lon='longitude',
                     hover_name='сity_name',
                     size='average_salary',
                     color='id',
                     labels={'id': 'Кол-во вакансий', 'average_salary': 'Средняя зарплата'},
                     projection='natural earth',
                     opacity = 0.8,
                     title='Географическое распределение вакансий по городам')

fig.update_geos(
    fitbounds="locations",
    showcoastlines=True
)
fig.show()



#### Вывод

Анализируя выполненную работу, мы замечаем следующие интересные моменты на графике: 
* Большие кружки представлены вне России, что может объясняться высокой ценностью этих вакансий или изменениями в курсе рубля на момент сбора данных с сайта HH.ru.
* Города с более крупным населением имеют более яркий цвет кружков, что указывает на большее количество представленных вакансий. Это наблюдается, например, в Москве, Санкт-Петербурге и Новосибирске.

In [58]:
connection.close()

# Общий вывод по проекту

> Вся работа была разбита на 7 блоков:
* Импорт основных библиотек для работы;
* знакомство с данными;
* предварительный анализ данных;
* детальный анализ вакансий;
* анализ работодателей;
* предметный анализ;
* дополнительные исследования. 

В первом блоке мы импортировали библиотеки, которые использовались в процессе работы для более корректного анализа данных;

Во втором блоке мы ознакомились с данными и узнали, что в них содержится 5 таблиц, которые использовались для анализа;

В третьем блоке мы провели предварительный анализ данных, включая распределение вакансий по регионам и оценку зарплатных интервалов. В этом блоке мы узнали что в таблице "Employers" содержится 23501 строк с наименованием работадателей и "Vacancies" в которой содержиться 49197 вакансий (строк);

В-четвертом блоке, постарались сделать первчиный анализ данных и выявили распределение вакансий по регионам, увидели размер средней минимальной и максимальной вилки по зарплате;

В пятом блоке сфокусировались на анализе работодателей и выявили интересные моменты в данных, такие как количество компаний, не указавших свою сферу деятельности а также выявили, что больше всего вакансий разместила компания "Яндекс" - 1933 вакансии в 181 регионах. Далее усложнив задачу, выписали 16 городов миллионников России с помощью запроса в "Wikipidia" и получили 485 вакансий, где в топ вошли такие города как Москва с 54 вакансиями и Санкт-Петербург с 42 вакансиями. 

В шестом блоке предметно расмотрели специалиста DataScience и результат оказался следующим, всего пристутсвует 51 вакансия для начинащих специалистов. Вакансий с требованием знания "Python", оказалось 351 шт. Вакансий с наличием навыка "SQL" и "Postgres" оказалось 201 шт. В среднем каждый DS специалист должен обладать как минимум 6 навыками. Специалисты у которых есть опыт от 3 до 6 лет в среднем могут получать 243115 руб.;

В-седьмом блоке, был подробно разобран специалист DS и его возможности. Были выгружен запрост с SQL, после чего создана модель для расчета средней зарплаты по каждому представленному региону. Далее выгружен с HTML страницы долгота и широта каждого региона, после чего был построен график карты, на котором отчетливо видно, что DS может получать высокую зарплату трудоустроившись в заграничную компанию, а также видно, что больше всего вакансий предложено в наших двух столицах. 

Работа проделана великолепно, и анализ данных может быть полезен для агентств по подбору вакансий и соискателей..  
 